packages/@lwc/engine-core/src/framework/modules/attr-unless-prop.ts

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright (c) 2018, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+import {
+    htmlAttributeToProperty,
+    isNull,
+    isUndefined,
+    StringCharCodeAt,
+    XML_NAMESPACE,
+    XLINK_NAMESPACE,
+} from '@lwc/shared';
+
+import { unlockAttribute, lockAttribute } from '../attributes';
+import { RendererAPI } from '../renderer';
+import { EmptyObject } from '../utils';
+import { VBaseElement } from '../vnodes';
+
+const ColonCharCode = 58;
+
+export function patchAttrUnlessProp(
+    oldVnode: VBaseElement | null,
+    vnode: VBaseElement,
+    renderer: RendererAPI
+) {
+    const {
+        data: { attrs },
+        elm,
+    } = vnode;
+    if (isUndefined(elm) || isUndefined(attrs)) {
+        return;
+    }
+
+    const { removeAttribute, setAttribute } = renderer;
+    const oldAttrs = isNull(oldVnode) ? EmptyObject : oldVnode.data.attrs;
+
+    for (const name in attrs) {
+        const cur = attrs[name];
+        const old = oldAttrs[name];
+
+        const propName = htmlAttributeToProperty(name);
+        if (propName in elm) {
```

Comment:
why not move this one check to attrs.ts and do the prop assignment there?

```
// attrs.ts
if(vnode.data.external && htmlAttributeToProperty(name) in elm) {
  (elm as any)[propName] = cur;
}
```
this will prevent duplication of the attribute assignment logic below.

Reply:
I don't think we want to slow down the regular cases with more conditions @abdulsattar, I think it is fine to have this small duplication.

Reply:
@caridy I did think of that but I realized that we're doing the check anyway: https://github.com/salesforce/lwc/pull/3150/files#diff-8622743d79e146365dd8c59fb85057a09b0f9577d99f065628abdd95d0d7a754R549

So, the first conditional in the above `if(vnode.data.external && ...)` is equivalent to what we have here.

Reply:
```js
if (propName in elm) {
```

Will this be an issue if `propName` is `constructor`? Or `__proto__`? Should this be a `hasOwnProperty` check instead?

Reply:
It can't be hasOwnProperty because inheritance on web components. but since our compiler is responsible for constructing the `attrs` dictionary per vnode, we can ban those names when parsing the element. It is probably already a problem with spread directive, so we better do that soon.

Reply:
@abdulsattar The check that you pointed out is done once per element, whereas the check required in attrs.ts would be once per attribute.

Reply:
For what it's worth [Vue](https://github.com/vuejs/core/blob/f3e4f038bf299794b2ca4750b4678074660113c8/packages/runtime-dom/src/patchProp.ts#L113) and [Svelte](https://github.com/sveltejs/svelte/blob/91f8764145f2c2b3c30bfec5a3b409df8e0ddcc9/src/runtime/internal/dom.ts#L328) uses the same `in` check to differentiate between props and attributes.